[3-4단계] 율무 - 방탈출 예약 대기 미션 제출

kkiseug opened the pull request:
안녕하세요! 샐리!!! 이번 리뷰도 잘 부탁드립니다.

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [방예univ](https://prolog.techcourse.co.kr/studylogs/4377)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
```text
[ ] 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.
[x] else 예약어를 쓰지 않는다.
[ ] 모든 원시 값과 문자열을 포장한다.
[ ] 일급 컬렉션을 쓴다.
[ ] 한 줄에 점을 하나만 찍는다.
[x] 줄여 쓰지 않는다(축약 금지).
[ ] 모든 엔티티를 작게 유지한다.
[ ] 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
[ ] getter/setter/프로퍼티를 쓰지 않는다.
```

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->
### Reservation과 WaitingReservation의 분리
저는 예약과 대기 예약을 두 개의 엔티티로 분리했습니다. 분리하는 과정이 순탄치는 않았어요. 처음엔 예약 자체의 책임에 집중해보았습니다. 예약이 갖고 있는 책임은 본인이 과거인지 판단하는 것 말고는 없어요. 사실 예약이 협력을 위한 객체라기보단 데이터적인 성향이 짙죠. 그렇다면, 정말 대기하는 예약과 책임이 다를까? 더 나아가 역할은?

저는 같다고 봤습니다. 하지만, DB에서 사용될 때는 다르게 봐야했죠. 왜냐면 '순서'라는 개념이 대기 예약에는 존재했기 때문입니다. 하지만 저는 순서라는 개념을 대기 예약 객체 자체가 가져야 할 책임보다는 오히려 서비스와 엔티티의 책임이라 생각했어요. 객체 자체가 순서를 갖고 있다는게 상태로 표현되는 책임 같았기 때문입니다.

그래서 처음엔 순수 Domain 객체 Reservation을 두고, Entity Reservation, WaitingReservation 두 개로 나누어 설계를 진행했습니다. 그러면, 순서라는 개념은 WaitingReservation 엔티티에만 둘 수 있고 서로의 책임은 같이 가져갈 수 있었기 때문이에요. 얼핏보면 괜찮은 생각이지만 사실 여기엔 크나큰 모순이 있었어요. 

그 모순은 순서라는 개념을 객체 자체의 책임으로 보지 않는다는 점입니다. 왜냐면, 엔티티는 언제든 변경될 수 있으며, 순서를 Service 혹은 일급 컬렉션이 순서를 보장한다는 것이 계속 변화하는 엔티티에 맞지 않고 일급 컬렉션의 경우 계층이 하나 더 늘어나는 셈이기 때문이에요! 오히려 개념을 객체 자체의 책임으로 보는게 객체지향에서 말하는 "프로그램의 복잡성을 줄이기 위한 추상화"라고 생각합니다.

그리고, Admin과 Member를 나눌 때 Reservation이 Admin까지 참조하는 꼴이 객체지향스럽지 않다고 생각했기에 Reservation도 같은 관점을 적용할 수 있었어요. 예를 들어 결제라는 요구사항이 들어오면 Reservation만 참조하는게 객체지향스러울텐데 분리하지 않으면 대기 예약도 참조하는 꼴이 되어버립니다.

그래서 분리의 개념을 도메인 객체까지 확장시켰고 지금의 구조가 나왔습니다.

### Lombok 적용
Lombok을 이번에 적용해보았어요. 처음 사용해봐서 제대로 사용하고 있는건지 모르겠습니다. 이 부분도 함께 봐주시면 좋겠어요!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

kkiseug committed:
feat: Lombok 의존성 추가 및 엔티티 적용

- 롬복을 통한 보일러 플레이트 코드 삭제

kkiseug committed:
feat: 예약 대기 프론트 기능 추가

- '예약대기' 버튼 추가 및 관련 이벤트 핸들러 onWaitButtonClick 구현
- 이미 예약된 시간에는 '예약대기' 버튼 활성화
- 예약 대기 생성 요청 로직 추가 기반 구조 작성
- 불필요한 'My Reservation' 메뉴 제거

kkiseug committed:
feat: 예약 대기 기능 추가

- `예약 대기` 생성을 위한 API 및 서비스 로직 추가
- `ReservationStatus`에 `WAITING` 상태 추가
- `JpaRepository`와 어댑터에 대기 예약 관련 쿼리 메서드 추가
- 대기 예약을 위한 `WaitingReservationEntity` 엔티티 생성
- 클라이언트 요청 시 예약 대기 API 호출 기능 추가

kkiseug committed:
refactor: 예약 대기 객체 분리 및 관련 로직 수정

- WaitingReservation 및 관련 DTO, Repository, Service 추가
- ReservationStatus 제거 및 관련 코드 수정
- 예약 대기 기능 컨트롤러 경로 및 로직 개선

kkiseug committed:
fix: 잘못된 reservation 테이블 쿼리 수정

- reservation 테이블에 존재하지 않는 reservation_status 컬럼 제거
- 기존 데이터 입력 쿼리 정상화

kkiseug committed:
chore: 예약 대기 생성 요청 API 호출 주석 제거

- 불필요한 TODO 주석 제거로 코드 정리

kkiseug committed:
refactor: 예약 관련 테스트 코드 수정

- `WaitingReservationQueryService` 및 관련 Repository, Bean 설정 추가
- `ReservationStatus.RESERVED` 제거로 인한 테스트 코드 수정

kkiseug committed:
feat: 예약 대기 삭제 기능 구현

- 관리자와 사용자 예약 삭제 API 분리 및 경로 수정
- 대기 예약 삭제 기능 추가 및 관련 유효성 검증 로직 구현
- MyReservationResponse 패키지 이동 및 필드 명 수정

kkiseug committed:
feat: 예약 삭제 시 인증 추가

- 테스트 케이스에 로그인 인증 추가 (토큰 생성 및 헤더에 포함)

kkiseug committed:
feat: 회원 데이터 초기화 SQL 스크립트 추가

- 테스트용 회원 데이터 insert 쿼리 추가

kkiseug committed:
refactor: 대기 예약 관련 패키지 분리 및 이동

- WaitingReservation 관련 도메인, 서비스, 인프라 코드를 reservation.waiting 패키지로 이동

kkiseug committed:
feat: 예약 대기 중복 확인 기능 추가

- 특정 테마, 시간, 날짜, 멤버 ID 기준으로 예약 대기 중복 여부 확인 로직 추가

kkiseug committed:
feat: 어드민 예약 대기 관리 페이지 기능 추가

kkiseug committed:
feat: 예약 대기 관리 페이지 기능 추가

- 예약 대기 목록 조회 API 및 승인/거절 기능 구현
- 관리자 전용 경로(/admin)로 수정
- waiting.html의 불필요한 내 예약 경로 제거

kkiseug committed:
feat: 예약 대기 관리 기능 추가

- 대기 예약 리스트 조회, 승인, 거절 API 구현
- ReservationResponse에 WaitingReservation 관련 메서드 추가

kkiseug committed:
refactor: 대기 예약 관련 로직 컨트롤러를 별도 클래스로 분리

kkiseug commented on src/main/java/roomescape/admin/domain/Admin.java:
```diff
@@ -0,0 +1,35 @@
+        this.name = name;
+        this.email = email;
+        this.password = password;
+    }
```
@CodeRabbit

1번은 의도를 나타내기 위함이고,2번도 어드민의 회원가입을 막고 있기때문에 의도적으로 검증하지 않았어.

3번은 엔티티인 이상 불변성을 보장하는게 더 이상한 거 같은데 어떻게 생각해?

kkiseug replied:
@CodeRabbit

Member는 회원가입을 할 수 있어서 원시값을 포장해서 검증하고 있어. 추후에 Admin에 대한 로직도 이런 방식으로 처리할 예정이야.

kkiseug commented on src/main/java/roomescape/theme/infrastructure/JpaThemeRepository.java:
```diff
@@ -0,0 +1,25 @@
+        ORDER BY COUNT(r) DESC
+        LIMIT :limit
+    """)
+    List<Theme> findPopularThemes(@Param("start_date") LocalDate start, @Param("end_date") LocalDate end, @Param("limit") int limit);
```
@CodeRabbit

Limit 쿼리는 몰랐네 고마워

kkiseug commented on src/main/java/roomescape/reservation/time/presentation/ReservationTimeController.java:
```diff
@@ -0,0 +1,68 @@
+        return ResponseEntity.ok(response);
+    }
+
+    @GetMapping(consumes = {"application/json"})
+    public ResponseEntity<List<TimeConditionResponse>> getReservationTimes(final TimeConditionRequest request) {
+        List<TimeConditionResponse> responses = reservationTimeService.getTimesWithCondition(request);
+        return ResponseEntity.ok().body(responses);
+    }
```
@CodeRabbit
오히려 명확하지 않아?

kkiseug committed:
refactor: JPQL limit절 제거

- Pageable로 대체

kkiseug committed:
refactor: 관리자 페이지 script 수정

kkiseug committed:
refactor: 예외 처리 메서드 네이밍 수정

kkiseug committed:
refactor: Exception 예외 응답코드 변경

- 예상치 못한 Exception이므로 500으로 반환

kkiseug committed:
feat: 회원과 관리자 예약 시간 API 분리

- 관리자 전용 예약 시간 API '/admin/times'로 변경
- 멤버 예약 시간 관련 컨트롤러 분리

kkiseug committed:
chore: 테스트용 어드민 계정 초기 데이터 추가

kkiseug committed:
test: 예약 대기 기능 테스트 추가

- WaitingReservationFacadeServiceTest 테스트 클래스 추가
- 대기 예약 생성, 중복 대기 방지, 승인 및 거절 등 조건별 테스트 작성

kkiseug committed:
refactor: 생성 시간 LocalDateTime으로 변경

kkiseug committed:
test: 관리자 예약 시간 API 테스트 수정

- 테스트 전 로그인 작업 추가 및 토큰 설정 메서드 추가
- 모든 테스트 요청에 인증 토큰 헤더 추가
- 기존 API 경로를 '/admin' prefix가 포함된 경로로 수정

kkiseug committed:
refactor: Facade 클래스명을 Application으로 변경

- 모든 Facade 클래스와 관련된 이름을 Application으로 변경

kkiseug committed:
refactor: 로그인 메서드 네이밍 수정

- createAdminToken, createMemberToken 메서드를 loginAdmin, loginMember로 네이밍 변경

kkiseug committed:
test: LoginService 테스트 코드 추가

- 어드민 및 멤버 로그인 성공/실패 케이스에 대한 테스트 작성
- 로그인 실패 시 적절한 예외 발생을 검증
- 테스트 실행을 위한 필요한 Bean 및 설정 추가

kkiseug committed:
refactor: 예약 대기 API 수정

kkiseug committed:
refactor: API 엔드포인트 URL 통일 및 개선

- `/reservations`를 `/reservation`로 변경하여 일관성 유지
- `/reservation-waiting`를 `/waiting-reservation`로 수정

kkiseug committed:
remove: 쓰지않는 프론트 파일 제거

kkiseug committed:
refactor: API 경로와 관련 로직 정리 및 수정

kkiseug committed:
refactor: 어드민 페이지 경로 모두 `/admin` 통하도록 변경

kkiseug committed:
feat: 로그아웃 기능 구현

- 로그아웃 시 토큰 즉시 삭제
- 로그아웃 후 로그인 페이지로 이동하도록 함

kkiseug committed:
test: 어드민 관련 페이지 접근 테스트 추가

kkiseug committed:
refactor: 예외 클래스 명칭 변경

- MemberNotFound -> MemberNotFoundException

kkiseug commented on src/main/java/roomescape/reservation/waiting/infrastructure/JpaWaitingReservationRepository.java:
```diff
@@ -0,0 +1,35 @@
+        FROM WaitingReservation w
+        WHERE w.member.id = :memberId
+        """)
+    List<WaitingReservationWithRank> findWaitingsWithRankByMember_Id(Long memberId);
```
이전 PR에서 Repository에 대한 제 생각을 말씀드렸는데, 그 생각에 따르면 Repository는 오직 도메인 객체에 대한 컬렉션처럼 동작해야해요. 하지만.. 이번 요구사항에 의해서 DTO를 반환하는 일이 생겼습니다. 제가 아직 DDD의 철학을 공부한 것이 아니라서 이런 경우가 괜찮은지 모르겠습니다.

dusdn1702 replied:
rank를 도메인으로 보지 않는 이유가 궁금해요. 저장하는 개체가 아닌 db에서 만들어오는 값이기 때문일까요?

kkiseug replied:
뭔가 rank를 설계하는 과정에서 repository에 책임을 넘겨주다보니까 자연스럽게 DB에 의존적인 개체로 생각했던 것 같습니다. 사실 처음 생각했던 구조로 갔으면 오히려 rank를 그런 시선으로 보지 않았을 거 같긴합니다. 다만.. 모르겠습니다. 도메인이란 정의를 아직 내리지 못했기 때문인지 rank를 어떤식으로 표현해야 할지 궁금하네요.

dusdn1702 replied:
db에서 가져오는 값이지만 waiting 이라는 대기 도메인에 순서가 중요한 책임을 가지고 있기 때문에 도메인에 포함시켜도 된다고 보여요.
도메인에 대해 광범위하고 어렵게 생각하기 보단, 프로젝트에서 핵심적인 역할을 하는 객체로 생각해봐도 좋겠어요.

kkiseug reviewed (commented):
# 궁금한 점

kkiseug commented:
@CodeRabbit

전체 코드를 리뷰하고 전체 변경 사항에 대해서 종합적인 평가도 남겨줘.
그리고, 코드 리뷰할 때 테스트에 대해서도 리뷰해주길 바랄게.

kkiseug commented:
@CodeRabbit

코드 코멘트 남겨줘.

dusdn1702 commented on build.gradle:
```diff
@@ -22,6 +22,9 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
     implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
 
+    compileOnly 'org.projectlombok:lombok'
```
lombok dependency를 이렇게 주입한 이유가 있으신가요?

kkiseug replied:
찾아보니까 gradle 버전에 따라 lombok 추가하는 방법이 다르더라고요 😓 그것도 모르고 컴파일 시점에만 필요한 것이라 생각해 아무 생각없이 compileOnly 썼네요. gradle 6.x 이상부터는 다른 방법이더라구요..

```text
implementation 'org.projectlombok:lombok'
annotationProcessor 'org.projectlombok:lombok'
```

변경했습니다!

dusdn1702 replied:
좋습니다~ 시간이 될 때 compileOnly의 동작 차이, gradle 선언의 동작 차이에 대해서 학습해두면 좋을 것 같아요.

dusdn1702 commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -5,8 +5,15 @@
 @Entity
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Getter
+@EqualsAndHashCode(of = "id")
```
필요한 어노테이션인가요?

kkiseug replied:
엔티티는 기본적으로 id값을 식별값으로 사용하기 때문에 VO처럼 모든 필드를 비교하는 것이 아닌 id만 비교해야 한다고 생각합니다. 그래서 Lombok을 통해 equals와 hashcode를 재정의하여 id로 비교하도록 만들었습니다.

dusdn1702 replied:
id만을 식별자로 사용하는 건 좋은 아이디어 같아요. 다만 불필요한 상황에서 어노테이션을 추가해둘 필요가 있을지는 검토해봐도 좋겠어요.

dusdn1702 commented on src/main/java/roomescape/reservation/application/ReservationApplicationService.java:
```diff
@@ -92,22 +83,41 @@ public List<ReservationResponse> searchReservationWithCondition(final SearchCond
+            .toList());
+
+        responses.addAll(waitings.stream()
+            .map(waiting -> new MyReservationResponse(
```
from을 사용하지 않는 이유가 있나요?

kkiseug replied:
없습니다.. 급하게 구현하느라 생각하지 못했던 것 같습니다. from으로 통일 했어요.

dusdn1702 commented on src/main/java/roomescape/reservation/presentation/MemberReservationController.java:
```diff
@@ -13,25 +13,23 @@
-        @RequestBody final ReservationRequest request,
-        @LoginMember final LoginMemberInfo memberInfo)
-    {
+    @PostMapping("/reservation")
```
reservations -> reservation 으로 변경하신 이유는 어떤 관점인지 궁금해요.

kkiseug replied:
하나의 예약과 관련된 api는 단수로 적용해주었습니다. 하나의 예약에 대해서 복수로 사용하는게 어색하다고 생각했기 때문입니다. API의 통일성을 해치는 것 같기도 하지만, 어느정도 예상 가능한 일관성이라고 생각합니다. 따로 rest API 설계 방식을 깊게 학습하진 않아서 제 기준에서 나름 적용해보았습니다.

샐리가 보기에는 어떤가요?

dusdn1702 replied:
restful 한 설계라고 많이들 얘기하는데 한번쯤 문서 읽어보고 율무의 기준을 잡아도 좋을 것 같아요. post라서 id가 pathVariable로 안붙어서 어색한 것이 맞기도 하고, 정답은 없는 법이니까요! 👍

dusdn1702 commented on src/main/java/roomescape/reservation/waiting/application/WaitingReservationApplicationService.java:
```diff
@@ -0,0 +1,114 @@
+package roomescape.reservation.waiting.application;
```
도메인을 예약/대기를 서로 다르게 보고 나누어서 구현해주셨는데 하나의 reservation 패키지에 두신 이유는 무엇인가요?

kkiseug replied:
일단 저는 기본적으로 예약/대기는 같은 '예약'이라는 도메인으로 묶인다고 생각합니다. 오히려 대기는 책임보다는 상태에 가깝죠. 하지만 순서라는 책임은 오히려 대기만 갖고 있어야 합니다. 그래서 처음에는 도메인 객체를 하나로 유지하고, 엔티티를 두 개로 나눈 뒤에 순서에 대한 개념을 일급 컬렉션으로 다루고 싶었어요.

<img width="568" alt="스크린샷 2025-05-27 10 52 27" src="https://github.com/user-attachments/assets/7c3d939f-904e-4c28-bdc9-97749b8b1245" />

그림으로 나타내면 이런 느낌이 될 거 같아요. WaitingService와 같은 서비스 단에서 일급 컬렉션을 활용해 WaitingReservation에 대한 것들을 수행하는거죠! 하지만, 이미 엔티티와 도메인의 개념을 통합해서 사용하고 있고 어느정도 프로그램의 복잡성을 줄이기 위한 관점에서도 서로 분리하는 현재의 구조가 더 낫다고 생각했어요. 다만 제 생각의 잔재를 패키지 구조로나마 표현해보았습니다..

dusdn1702 replied:
좋습니다 이해했어요 ㅎㅎ  
다만, waiting 도메인을 reservation 하나로 합쳐서 표현해보는 것도 좋을 것 같은데, 시간이 부족하다면 이 상태에서 여유가 될 때 디벨롭해봐도 좋겠어요.  
구현 없이 패키지 구조로 표현되기엔 쉽게 이해하기엔 다른 개발자들 입장에선 어려울 수 있다고 생각해요!

dusdn1702 commented on src/main/java/roomescape/reservation/waiting/domain/WaitingReservation.java:
```diff
@@ -0,0 +1,81 @@
+        return date.isEqual(dateTime.toLocalDate()) && time.isBefore(dateTime.toLocalTime());
+    }
+
+    public boolean isSameTime(final ReservationTime time) {
```
사용되지 않는 메서드네요~

kkiseug replied:
맞네요!!! 수정하였습니다.

dusdn1702 commented on src/main/java/roomescape/reservation/waiting/infrastructure/JpaWaitingReservationRepository.java:
```diff
@@ -0,0 +1,35 @@
+    @Query("""
+        SELECT new roomescape.reservation.waiting.domain.dto.WaitingReservationWithRank(
+            w,
+            (SELECT COUNT(w2)
```
rank를 계산하는 책임을 repository에서 구현해주신 이유가 궁금해요. pr에서 말씀 주셨던 책임의 관점에서 더 이야기해보면 좋겠어요.

dusdn1702 replied:
추가로 이렇게 되면 예약과 대기가 분리되어야 하는 이유가 무엇인지도 더 자세히 설명해주실 수 있으면 좋겠어요!

kkiseug replied:
위의 답변처럼 rank의 책임을 처음엔 일급 컬렉션으로 가져가려고 했는데, 엔티티와 도메인을 통합한 순간부터 이를 누군가의 책임으로 주려면 도메인에 대한 컬렉션인 Repository 밖에 없다고 생각했어요. 그래서 repository에 넘겨주었습니다. 그래서 저는 오히려 둘이 분리되는 것이 자연스럽게 느껴지긴 합니다!!

dusdn1702 replied:
넵 위의 답변에서 이해한 내용이라 넘어갈게요 👍

dusdn1702 commented on src/main/java/roomescape/reservation/application/ReservationApplicationService.java:
```diff
@@ -92,22 +83,41 @@ public List<ReservationResponse> searchReservationWithCondition(final SearchCond
+                waiting.waitingReservation().getTheme().getName(),
+                waiting.waitingReservation().getDate(),
+                waiting.waitingReservation().getTime().getStartAt(),
+                (waiting.rank()+1) + "번째 예약대기"
```
status에 대해서 service에서 노출을 위한 문구를 만들어주고 있는데 다른 방법으로는 어떤 것이 있을까요?

kkiseug replied:
DTO 내에서 from 메서드로 통합하면서 해당 DTO내에 넣어주었습니다. status의 노출이 결국 View를 위한 것이기 때문에 이런 흐름이 더 자연스러운 것 같습니다.

dusdn1702 replied:
좋습니다! 👍

dusdn1702 commented on src/main/java/roomescape/reservation/waiting/domain/WaitingReservation.java:
```diff
@@ -0,0 +1,81 @@
+    @ManyToOne
+    private Member member;
+
+    @CreationTimestamp
```
어떤 어노테이션인가요? 사용자가 저장한 시간이 아닌 db에 저장된 시간을 사용하는 이유가 있나요?

kkiseug replied:
JPA가 DB에 엔티티를 저장할 때 현재 시간을 기준으로 자동으로 해당 컬럼을 채워 쿼리를 만들어주는 애노테이션입니다! Hibernate 구현체에 종속된 애노테이션이더라구요. 그래서 다른 애노테이션이 있나봤는데, 스프링에서 제공하는 `@CreatedDate` 애노테이션이 있더라구요. 특정 구현체의 애노테이션에 종속되는 것이 조금 별로인 형태긴 하나 JPA가 결국 Hibernate 구현체를 쓰는 경우가 대다수인점을 생각하면 설정이 좀 더 필요한 `@CreatedDate` 애노테이션보다는 Hibernate 애노테이션을 써도 될 것 같은데 샐리는 어떻게 생각하시나요?

그리고 처음엔 DB에 저장된 시간을 기점으로 저장하는 것이 뭔가 더 일관성 있고 형평성이 있다 생각했는데 지금은 조금 모호하네요. 서비스에서 이를 지정해주는 방식이 서비스 관점에서 더 알맞는 거 같기도 합니다. DB 종속성인 문제를 피할 수도 있구요. 테스트하기도 더 쉬울 것 같기도 해요. 어떤 요구사항이냐에 따라 다를 거 같긴한데, 보통은 어떤 방식을 선호하시나요?

dusdn1702 replied:
하나의 기준을 잡고 가면 좋을 것 같아요.  
지금은 등록된 시간에 맞게 대기 순서를 부여하기 때문에 DB에 종속적이라 저장 시간도 생성 시간으로 보는게 일관성 있어 보여요. 그 과정에서 어노테이션의 종류는크게 중요하지 않다고 생각하는데 다른 컬럼들에서 사용하는 스프링 어노테이션이 낫지 않을까요?   
일반적으로 저는 개인적으로 서버에서 이 시간을 관리하는 편이에요.

dusdn1702 commented on src/main/java/roomescape/reservation/waiting/infrastructure/JpaWaitingReservationRepositoryAdaptor.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @Override
+    public List<WaitingReservationWithRank> findWaitingsWithRankByMember_Id(Long memberId) {
```
adaptor에서 다른 메서드들과 아예 다른 메서드명인데요. 
명명 규칙을 어떻게 갖고 계신가요?

kkiseug replied:
급해서 일단 LMS에서 제공하는 이름으로 결정했던 거 같아요. 좀 더 비즈니스적인 이름을 담아야 했다고 생각합니다! 근데 이번에 느낀게 확실히 계층이 분리되다보니 급하면 좀 헷갈리긴 히더라고요 ... ㅋㅋㅋ

dusdn1702 replied:
넵 adaptor의 단점이 여기서 부각될 수 있겠네요 ㅎㅎ  
지금은 경험상 적용해봤다고 생각하시되 여기서 분리했을 때의 장단점은 잘 정리해두시면 좋겠어요!

dusdn1702 commented on src/main/java/roomescape/reservation/waiting/infrastructure/JpaWaitingReservationRepository.java:
```diff
@@ -0,0 +1,35 @@
+
+public interface JpaWaitingReservationRepository extends CrudRepository<WaitingReservation, Long> {
+
+    @Transactional
```
트랜잭션을 repository에서 일부에만 구현하신 이유가 무엇인가요? 
원하는 동작과 실제 동작이 어떻게 되는지 같이 한번 설명해주실 수 있을까요?

kkiseug replied:
분명 Transactional을 붙이지 않았을 때 에러가 발생했었는데.. 지금은 또 안나네요.. 무슨 동작을 의도하고 붙인 것은 아니고 구현 당시에 애노테이션을 붙이지 않았을 때는 빨간줄이 뜨다가 혹시나 해서 붙여보니 에러가 발생하지 않았던 기억이 납니다. 그래서 그때는 SimpleJpaRepository 구현체의 문제라고 생각했는데 이번에 지우니까 또 잘되네요..

네.. 그래서 어떤 의도가 있었던 부분은 아닙니다!!!

dusdn1702 replied:
ㅋㅋㅋㅋ 빨간 줄은 모르겠지만..  
Transactional을 어디에 붙일지도 학습해보시면 좋겠어요.  
예약 대기를 하고 승인/거부가 되었는데 누군가는 예약이 되어버리면 이상하겠죠?? 이걸 어떻게 막을 수 있을지 생각해보면 좋겠어요.

dusdn1702 reviewed (changes requested):
안녕하세요 율무!
2단계 미션도 잘 구현해주셨네요. 고생 많으셨습니다.
새로 구현된 WaitingReservation 위주로 코멘트 남겨보았어요.
확인 부탁드릴게요!

kkiseug committed:
feat: 테스트 환경에서는 test에서 정의한 data.sql을 읽도록 설정

kkiseug committed:
refactor: 서비스 테스트에서 @DataJpaTest 애노테이션 제거

- @SpringBootTest에서 옵션으로 필요한 빈만 생성하도록 함
- Config 파일의 제거
- DB 롤백을 위한 클래스 추가

kkiseug committed:
refactor: lombok 의존성 주입 방식 변경

kkiseug committed:
refactor: MyReservationResponse 생성 로직 통합

kkiseug committed:
refactor: 사용하지 않는 메서드 제거

kkiseug committed:
refactor: 메서드 명명 개선으로 가독성 향상

- findWaitingsWithRankByMember_Id를 findWaitingReservationsWithRank로 변경
- existsByThemeIdAndTimeIdAndDateAndMemberId를 isAlreadyWaiting으로 변경
- 메서드명을 더 직관적으로 표현하여 코드 가독성을 개선

kkiseug committed:
refactor: @Transactional 어노테이션 제거

- deleteByIdAndMemberId 메서드에서 불필요한 @Transactional 어노테이션 제거

kkiseug reviewed (commented):
리뷰 감사합니다~! 마지막에 설계 고민 + 느긋함 때문에 급하게 마무리 짓느라 의도를 많이 고민하지 못했던 것 같네요 ㅜㅜㅜ 토론하는 것을 좋아하는 성격이라 아쉽습니다.

dusdn1702 commented on src/main/java/roomescape/reservation/waiting/infrastructure/JpaWaitingReservationRepositoryAdaptor.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @Override
+    public boolean isAlreadyWaiting(Long themeId, Long timeId, LocalDate date, Long memberId) {
```
adaptor 네이밍을 수정해주셨는데요! adaptor에게 각 구현체에 의존적이지 않은 repository성 함수들을 주고 싶은지, service의 성격을 띄는 함수들을 주고 싶은지 생각해보면 좋겠어요.

dusdn1702 commented on src/test/java/roomescape/common/BaseServiceTest.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
+
+@SpringBootTest(webEnvironment = WebEnvironment.NONE)
+public abstract class BaseServiceTest {
```
테스트를 위한 설정을 분리해주셨군요! 👍

dusdn1702 reviewed (approved):
안녕하세요 율무!  
피드백 반영해주신 부분들과 코멘트들에 대한 답변 남겨두었어요.  
코드적으로 더 수정할 부분은 보이지 않아서 이번 미션은 여기서 마무리하도록 할게요.  
남긴 코멘트 중에 궁금하거나 더 이야기하고 싶은 부분 있으시면 편하게 말씀주세요!  
고생 많으셨습니다. 앞으로도 화이팅입니다~
